Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHPML.cpp

```diff
@@ -0,0 +1,162 @@
+/* Copyright 2020 Remi Lehe
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "Utils/WarpXAlgorithmSelection.H"
+#include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H"
+#ifdef WARPX_DIM_RZ
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H"
+#else
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianYeeAlgorithm.H"
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianCKCAlgorithm.H"
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianNodalAlgorithm.H"
+#endif
+#include "BoundaryConditions/PMLComponent.H"
+#include <AMReX_Gpu.H>
+#include <AMReX.H>
+#include "Utils/CoarsenIO.H"
+
+using namespace amrex;
+
+#ifdef WARPX_MAG_LLG
+
+/**
+ * \brief Update the H field, over one timestep
+ */
+void FiniteDifferenceSolver::MacroscopicEvolveHPML (
+    std::array< amrex::MultiFab*, 3 > Hfield,
+    std::array< amrex::MultiFab*, 3 > const Efield,
+    amrex::Real const dt,
+    std::unique_ptr<MacroscopicProperties> const& macroscopic_properties,
+    amrex::MultiFab* const mu_mf ) {
+
+   // Select algorithm (The choice of algorithm is a runtime option,
+   // but we compile code for each algorithm, using templates)
+#ifdef WARPX_DIM_RZ
+    amrex::ignore_unused(Hfield, Efield, dt, macroscopic_properties, mu_mf);
+    amrex::Abort("PML are not implemented in cylindrical geometry.");
+#else
+    if (m_do_nodal) {
+
+        MacroscopicEvolveHPMLCartesian <CartesianNodalAlgorithm> ( Hfield, Efield, dt, macroscopic_properties, mu_mf );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::Yee) {
+
+        MacroscopicEvolveHPMLCartesian <CartesianYeeAlgorithm> ( Hfield, Efield, dt, macroscopic_properties, mu_mf );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::CKC) {
+
+        MacroscopicEvolveHPMLCartesian <CartesianCKCAlgorithm> ( Hfield, Efield, dt, macroscopic_properties, mu_mf );
+
+    } else {
+        amrex::Abort("Unknown algorithm");
+    }
+#endif
+}
+
+
+#ifndef WARPX_DIM_RZ
+
+template<typename T_Algo>
+void FiniteDifferenceSolver::MacroscopicEvolveHPMLCartesian (
+    std::array< amrex::MultiFab*, 3 > Hfield,
+    std::array< amrex::MultiFab*, 3 > const Efield,
+    amrex::Real const dt,
+    std::unique_ptr<MacroscopicProperties> const& macroscopic_properties,
+    amrex::MultiFab* const mu_mf ) {
+
+    amrex::GpuArray<int, 3> const& mu_stag  = macroscopic_properties->mu_IndexType;
+    amrex::GpuArray<int, 3> const& Hx_stag  = macroscopic_properties->Hx_IndexType;
+    amrex::GpuArray<int, 3> const& Hy_stag  = macroscopic_properties->Hy_IndexType;
+    amrex::GpuArray<int, 3> const& Hz_stag  = macroscopic_properties->Hz_IndexType;
+    amrex::GpuArray<int, 3> const& macro_cr = macroscopic_properties->macro_cr_ratio;
```

Comment:
I noticed if these lines were put after Line 80, it wouldn't work. It would not detect `mu_stag` and return an error that `mu_stag is not declared in this scope`. 

Could you help me understand the reason? 

My straightforward understanding would be the `#ifdef _OPENMP` . I know with `#ifdef ` the kernel compilation could be complex (thank you Andy for explaining this ). But in this case I don't see the link... 

Thank you!

Reply:
I just tried this experiment and the code compiles and runs fine.  If you can replicate it again, send me the .cpp file.

Reply:
Thank you for the explanation!